src/main/java/org/opensearch/jobscheduler/utils/JobDetailsService.java

```diff
@@ -33,31 +33,89 @@
 import org.opensearch.common.xcontent.LoggingDeprecationHandler;
 import org.opensearch.index.IndexNotFoundException;
 import org.opensearch.index.engine.DocumentMissingException;
+import org.opensearch.index.engine.Engine;
 import org.opensearch.index.engine.VersionConflictEngineException;
 import org.opensearch.index.seqno.SequenceNumbers;
+import org.opensearch.index.shard.IndexingOperationListener;
+import org.opensearch.index.shard.ShardId;
 import org.opensearch.jobscheduler.model.JobDetails;
 
 import java.nio.charset.StandardCharsets;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 
-public class JobDetailsService {
+public class JobDetailsService implements IndexingOperationListener {
 
     private static final Logger logger = LogManager.getLogger(JobDetailsService.class);
-    private static final String JOB_DETAILS_INDEX_NAME = ".opensearch-plugins-job-details";
+    public static final String JOB_DETAILS_INDEX_NAME = ".opensearch-plugins-job-details";
     private static final String PLUGINS_JOB_DETAILS_MAPPING_FILE = "/mappings/opensearch_plugins_job_details.json";
 
     public static Long TIME_OUT_FOR_REQUEST = 10L;
     private final Client client;
     private final ClusterService clusterService;
+    private Set<String> indicesToListen;
 
-    public JobDetailsService(final Client client, final ClusterService clusterService) {
+    private static final ConcurrentMap<String, JobDetails> indexToJobDetails = IndexToJobDetails.getInstance();
+
+    public JobDetailsService(final Client client, final ClusterService clusterService, Set<String> indicesToListen) {
         this.client = client;
         this.clusterService = clusterService;
+        this.indicesToListen = indicesToListen;
     }
 
     public boolean jobDetailsIndexExist() {
         return clusterService.state().routingTable().hasIndex(JOB_DETAILS_INDEX_NAME);
     }
 
+    public static ConcurrentMap<String, JobDetails> getIndexToJobDetails() {
+        return JobDetailsService.indexToJobDetails;
+    }
+
+    private void updateIndicesToListen(String jobIndex) {
+        this.indicesToListen.add(jobIndex);
+    }
+
+    /**
+     * Registers or updates a jobDetails entry
+     *
+     * @param extensionId the unique Id for the job details
+     * @param jobDetails the jobDetails to register
+     */
+    void updateIndexToJobDetails(String extensionId, JobDetails jobDetails) {
+        if (indexToJobDetails.containsKey(extensionId)) {
```

Comment:
This whole set of conditionals is confusing.  It may be more readable by making the `getJobType()` null check the outer conditional, and using `getOrDefault()` or other map methods.  In any case, the javadoc should be more descriptive of what's happening in here "updates the map to include job details if X, or adds job details if Y"